cijoe/scripts/fio_latency.py

```diff
@@ -0,0 +1,404 @@
+class Engine:
+    name: str
+    device: Device
+    cijoe: Cijoe
+
+    @property
+    def name_id(self) -> str:
+        return f"{self.name}"
+
+    @property
+    def ioengine(self) -> str:
+        return self.name
+
+    def env(self) -> Dict[str, str]:
+        """Environment variables used in execution of test."""
+        return {}
+
+    @property
+    def filename(self) -> str:
+        return self.device.uri
+
+    @property
+    def direct(self) -> str:
+        """Value for the `--direct` fio flag."""
+        return "1"
+
+    def binary(self, bin: str) -> str:
+        """Path to fio binary"""
+        return bin
+
+    def extra_args(self) -> Dict[str, str]:
+        """Method that can be overridden to add extra arguments when running fio.
+        e.g. xnvme backend engine when it applies.
+        """
+        return {}
+
+    def prepare(self) -> None:
+        """Method that will run in preparation for execution fio test.
+        e.g. is used to setup nvme devices.
+        """
+
+
+class KernelEngine(Engine):
+    def prepare(self) -> None:
+        self.cijoe.run(" ".join(["xnvme-driver", "reset"]))
+
+
+class UserSpaceEngine(Engine):
+    def prepare(self) -> None:
+        self.cijoe.run(" ".join(["xnvme-driver"]))
+
+
+@dataclass
+class BuiltinEngine(KernelEngine):
+    ...
+
+
+@dataclass
+class XnvmeKernelEngine(KernelEngine):
+    be: str
+    name: str
+    device: Device
+    cijoe: Cijoe
+
+    @property
+    def name_id(self) -> str:
+        return f"{self.name}_{self.be}"
+
+    def extra_args(self) -> Dict[str, str]:
+        return {"-xnvme_be": "linux", "-xnvme_async": f"{self.be}"}
+
+
+@dataclass
+class XnvmeUserSpaceEngine(UserSpaceEngine):
+    be: str
+    name: str
+    device: Device
+    cijoe: Cijoe
+
+
+@dataclass
+class ExternalPreloader(UserSpaceEngine):
+    path: Path
+    fio_bin: Path
+    name: str
+    device: Device
+    cijoe: Cijoe
+
+    def env(self) -> Dict[str, str]:
+        return {"LD_PRELOAD": f"{self.path}"}
+
+    def binary(self, bin: str) -> str:
+        return f"{self.fio_bin}"
+
+    @property
+    def filename(self) -> str:
+        return f"trtype=PCIe traddr={self.device.uri.replace(':', '.')} ns=1"
+
+
+@dataclass
+class XnvmeNullEngine(XnvmeKernelEngine):
+    @property
+    def name_id(self) -> str:
+        return "xnvme_null"
+
+
+@dataclass
+class XnvmeSPDK(XnvmeUserSpaceEngine):
+    def extra_args(self) -> Dict[str, str]:
+        return {
+            "-xnvme_be": f"{self.be}",
+            "-xnvme_async": "nvme",
+            "-xnvme_dev_nsid": str(self.device.nsid),
+        }
+
+    @property
+    def name_id(self) -> str:
+        return "xnvme_spdk"
+
+    @property
+    def filename(self) -> str:
+        return self.device.uri.replace(":", "\:")  # noqa: W605
+
+
+@dataclass
+class XnvmeIOURingCMD(XnvmeKernelEngine):
+    @property
+    def direct(self) -> str:
+        return "0"
+
+
+@dataclass
+class IOURingCmdBuiltinEngine(KernelEngine):
```

Comment:
Perhaps the class name can be simplified, i.e. removing the reference to BuiltinEngine